tls/client-ech.c

```diff
@@ -0,0 +1,194 @@
+    wolfSSL_Init();
+    method = wolfTLSv1_3_client_method(); /* use TLS v1.3 */
+
+    /* make new ssl context */
+    if ((ctx = wolfSSL_CTX_new(method)) == NULL) {
+        ret = 1;
+        goto end;
+    }
+
+    /* set the server name we want to connect to */
+    ret = wolfSSL_CTX_UseSNI( ctx, WOLFSSL_SNI_HOST_NAME,
+        "crypto.cloudflare.com", strlen("crypto.cloudflare.com") );
+
+    if (ret != WOLFSSL_SUCCESS) {
+        printf("wolfSSL_CTX_UseSNI error %d", ret);
+        goto end;
+    }
+
+    /* make new wolfSSL struct */
+    if ((ssl = wolfSSL_new(ctx)) == NULL) {
+        printf("wolfSSL_new error");
+        ret = 1;
+        goto end;
+    }
+
+    /* Add cert to ctx */
+    if ((ret = wolfSSL_CTX_load_verify_locations(ctx, CERT_FILE, 0)) !=
+      WOLFSSL_SUCCESS) {
+        printf("wolfSSL_CTX_load_verify_locations error %d", ret);
+        goto end;
+    }
+
+    /* Connect wolfssl to the socket, server, then send message */
+    wolfSSL_set_fd(ssl, sockfd);
+
+    /* this connect will send a grease ech and get the retry configs back */
+    ret = wolfSSL_connect(ssl);
+
+    if (ret != WOLFSSL_SUCCESS) {
+        printf("%d %d\n", ret, wolfSSL_get_error( ssl, ret ));
+        goto end;
+    }
+
+    /* retrieve the retry configs sent by the server */
+    ret = wolfSSL_GetEchConfigs(ssl, ech_configs, &ech_configs_len);
+
+    if (ret != WOLFSSL_SUCCESS) {
+        printf("wolfSSL_GetEchConfigs error %d\n", ret);
+        goto end;
+    }
+
+    /* frees all data before client termination */
+    wolfSSL_free(ssl);
+    close(sockfd);
+
+    /* now we create a new connection that will send the real ech */
+
+    /* create and set up socket */
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    memset(&servAddr, 0, sizeof(servAddr));
+    servAddr.sin_family = AF_INET;
+    servAddr.sin_port = htons(SERV_PORT);
+    /* set the ip string to the cloudflare server */
+    servAddr.sin_addr.s_addr = inet_addr( ip_string );
+
+    /* connect to socket */
+    connect(sockfd, (struct sockaddr *) &servAddr, sizeof(servAddr));
+
+    /* make new wolfSSL struct */
+    if ( (ssl = wolfSSL_new(ctx)) == NULL) {
+        printf("wolfSSL_new error");
+        ret = 1;
+        goto end;
+    }
+
+    /* set the ech configs taken from dns */
+    ret = wolfSSL_SetEchConfigs(ssl, ech_configs, ech_configs_len);
+
+    if ( ret != WOLFSSL_SUCCESS ) {
+        printf("wolfSSL_SetEchConfigs error %d", ret);
+        goto end;
+    }
+
+    /* Connect wolfssl to the socket, server, then send message */
+    wolfSSL_set_fd(ssl, sockfd);
+
+    /* this connect will send the real ech */
+    ret = wolfSSL_connect(ssl);
+
+    if (ret != WOLFSSL_SUCCESS) {
+        printf( "%d %d\n", ret, wolfSSL_get_error( ssl, ret ) );
+        goto end;
+    }
+
+    wolfSSL_write(ssl, message, strlen(message));
+
+    do
+    {
+        ret = wolfSSL_read( ssl, rd_buf, sizeof(rd_buf) );
+        printf( "%.*s", ret, rd_buf );
```

Comment:
Read with rd_buf -1 and please null terminate since you are using in printf.

Reply:
`%.*s` takes the length and then the buffer so it doesn't need a NULL terminator